packages/charts/src/utils/data/date_time.ts

```diff
@@ -18,3 +19,85 @@ export function getMomentWithTz(date: number | Date, timeZone?: string) {
   }
   return moment.tz(date, timeZone);
 }
+
+/** @internal */
+export type UnixTimestamp = number;
+
+/** @internal */
+export function snapDateToInterval(
+  date: number | Date,
+  interval: 'calendar' | 'fixed',
+  unit: string,
+  snapTo: 'start' | 'end',
+  timeZone?: string,
+): UnixTimestamp {
+  const momentDate = getMomentWithTz(date, timeZone);
+
+  return interval === 'calendar'
+    ? calendarIntervalSnap(momentDate, unit, snapTo).valueOf()
+    : fixedIntervalSnap(momentDate, unit, snapTo).valueOf();
+}
+
+function calendarIntervalSnap(date: moment.Moment, unit: string, snapTo: 'start' | 'end') {
+  const momentUnitName = esCalendarIntervalsToMoment(unit);
+  if (!momentUnitName) {
+    return date;
+  }
+  return snapTo === 'start' ? date.startOf(momentUnitName) : date.endOf(momentUnitName);
+}
+function fixedIntervalSnap(date: moment.Moment, unit: string, snapTo: 'start' | 'end') {
+  const unitMultiplier = esFixedIntervalsToMomentUnit(unit);
+  if (isNaN(unitMultiplier)) {
+    return date;
+  }
+
+  const roundedDate = Math.floor(date.valueOf() / unitMultiplier) * unitMultiplier;
+  return snapTo === 'start' ? roundedDate : roundedDate + unitMultiplier - 1;
+}
+
+function esFixedIntervalsToMomentUnit(unit: string): number {
+  if (unit.endsWith('ms')) {
+    return getNumericalUnit(unit, 'ms');
+  }
+  if (unit.endsWith('s')) {
+    return getNumericalUnit(unit, 's') * 1000;
+  }
+  if (unit.endsWith('m')) {
+    return getNumericalUnit(unit, 'm') * 1000 * 60;
+  }
+  if (unit.endsWith('h')) {
+    return getNumericalUnit(unit, 'h') * 1000 * 60 * 60;
+  }
+  if (unit.endsWith('d')) {
+    return getNumericalUnit(unit, 'd') * 1000 * 60 * 60 * 24;
+  }
+  return NaN;
+}
+
+function getNumericalUnit(unit: string, intervalLabel: string) {
+  return Number.parseFloat(unit.slice(0, unit.indexOf(intervalLabel)));
+}
+
+function esCalendarIntervalsToMoment(unit: string): unitOfTime.StartOf | undefined {
+  if (unit === 'minute' || unit === '1m') {
+    return 'minutes';
+  }
+  if (unit === 'hour' || unit === '1h') {
+    return 'hour';
+  }
+  if (unit === 'day' || unit === '1d') {
+    return 'day';
+  }
+  if (unit === 'week' || unit === '1w') {
+    return 'week';
+  }
+  if (unit === 'month' || unit === '1M') {
+    return 'month';
+  }
+  if (unit === 'quarter' || unit === '1q') {
+    return 'quarter';
+  }
+  if (unit === 'year' || unit === '1y') {
+    return 'year';
+  }
+}
```

Comment:
Just a thought 💭 ...

```suggestion
const esMomentIntervalMapping = {
  minute: 'minutes', 
  '1m': 'minutes',
  hour: 'hour',
  '1h': 'hour',
  day: 'day',
  '1d': 'day',
  week: 'week',
  '1w': 'week',
  month: 'month',
  '1M': 'month',
  quarter: 'quarter',
  '1q': 'quarter',
  year: 'year',
  '1y': 'year',
}

function esCalendarIntervalsToMoment(unit: string): unitOfTime.StartOf | undefined {
  esMomentIntervalMapping[unit]
}
```

Reply:
Short, less code noise (curlies), no branching and O(1) 👍 

Would it be possible to collect time related utils in the monorepo but outside `charts`? Eg. under some `time-utils`. It's one of those things on which various bits of Kibana need to depend, and now, everybody implements their own thing.

Reply:
Also, the `unitOfTime.StartOf | undefined` union type may be avoided, see my other note

Reply:
@monfera
>Would it be possible to collect time related utils in the monorepo but outside charts? Eg. under some time-utils. It's one of those things on which various bits of Kibana need to depend, and now, everybody implements their own thing.

we could but we can't at the moment publish multiple packages easily due to the semantic-release tool.

Reply:
> @monfera
> 
> > Would it be possible to collect time related utils in the monorepo but outside charts? Eg. under some time-utils. It's one of those things on which various bits of Kibana need to depend, and now, everybody implements their own thing.
> 
> we could but we can't at the moment publish multiple packages easily due to the semantic-release tool.

Yeah I think we need to figure out how this would work or either as parallel packages published as one or multiple published packages consumed by charts. I do think though that we can publish libraries that are only development util functions like the random number generator that eui and others use, in this case the release notes of that package are not really important.

Or we could do it as react, storybook and others and just publish all packages under the same version and release notes.

Reply:
I've refactored the code [here](https://github.com/elastic/elastic-charts/pull/1253/commits/c93defff89c45bad4b0ceb5f6255e70f797224b2), using a switch in one case, and an object map in another